bpo-36889: Document Stream and StreamServer. (#14203)
python/cpython: Doc/library/asyncio-stream.rst

    import asyncio

    async def tcp_echo_client(message):
        async with asyncio.connect('127.0.0.1', 8888) as stream:
            print(f'Send: {message!r}')
            await stream.write(message.encode())

            data = await stream.read(100)
            print(f'Received: {data.decode()!r}')

>             print('Close the connection')

Reviewer: This line is confusing, it is not followed by an actual closing code.
Maybe better to drop it, `async with` provides enough implication that the stream is closed on exit from the block I guess

Author: Yes, the message about closing is not needed. Removed. Thanks.

    import asyncio

    async def tcp_echo_client(message):

>         stream = await asyncio.connect('127.0.0.1', 8888)

Reviewer: `connect` supports context manager:
```
async with asyncio.connect('127.0.0.1', 8888) as stream:
    await stream.write(...)
```
It can be more idiomatic that current version.
What do you think?

Author: +1 for context managers. I rewrote the examples and they are much more idiomatic and concise now.

        print(f'Send: {message!r}')

>         stream.write(message.encode())

Reviewer: ```suggestion
        await stream.write(message.encode())
```
Missing `await` is supported but discouraged.

Author: Thanks for the suggestion. Changed in this example and in others to make sure write calls are awaited.

and work with streams:

.. coroutinefunction:: connect(host=None, port=None, \*, \

>                                loop=None, limit=2**16, ssl=None, family=0, \

Reviewer: `loop` parameter is not supported

Author: Thanks, removed.

                               proto=0, flags=0, sock=None, local_addr=None, \
                               server_hostname=None, ssl_handshake_timeout=None, \
                               happy_eyeballs_delay=None, interleave=None)

   This function internally uses :meth:`loop.create_connection` to return an awaitable
   :class:`Stream` object of the mode ``READWRITE`` that can be used as a reader

>    and a writer. The function can also be used as an async context manager where it's

Reviewer: Just `The function can also be used as an async context manager`, 
drop `where it's automatically awaited`.
Technically it's not correct: `connect` returns an internal undocumented object that can be awaited or used by `async with`. I don't want to document this object but want to provide a clear vision of how to use `connect` call.

Maybe two short usage examples can help?
E.g.
```
Connect to TCP socket on *host* : *port* address, return a read-write :class:`Stream` object.

The function can be used with ``await`` to get a connected stream::
    stream = await asyncio.connect('127.0.0.1', 8888)

Also, ``async with`` form is supported, the stream is closed on exit from the block::

   async with asyncio.connect('127.0.0.1', 8888) as stream:
        ...

   *limit* determines the buffer size limit used by the returned :class:`Stream` instance. By default the limit is set to 64 KiB.

    The rest of the arguments are passed directly to :meth:`loop.create_connection`.

```

I'm not sure about the exact text but I hope you've got my idea. Documentation is hard.

Author: Agreed it's hard :) I have adopted your suggestion and formatted it. Did the same for `connect_unix` too.

and work with streams:

.. coroutinefunction:: connect(host=None, port=None, \*, \
                               loop=None, limit=None, ssl=None, family=0, \
                               proto=0, flags=0, sock=None, local_addr=None, \
                               server_hostname=None, ssl_handshake_timeout=None, \
                               happy_eyeballs_delay=None, interleave=None)

   This function internally uses :meth:`loop.create_connection` to return a
   :class:`Stream` object of the mode ``READWRITE`` that can be used as a reader

>    and a writer.

Reviewer: Please reflect the fact that `async with connect()`  can be used as well as just `await connect()`.

Author: I have added a note that connect returns a stream object which is awaitable and also about using it as context manager. Let me know if it needs to be reworded that connect itself is awaitable.

   .. deprecated-removed:: 3.8 3.10

      `start_server()` is deprecated if favor of :class:`StreamServer`

.. coroutinefunction:: connect_read_pipe(pipe, *, limit=2**16)

>    Takes a :term:`file-like object <file object>` to return a

Reviewer: ```suggestion
   Takes a :term:`file-like object <file object>` *pipe* to return a
```

   :class:`Stream` object of the mode ``READ`` that has similar API of
   :class:`StreamReader`. It can also be used as an async context manager.

> 

Reviewer: ```
   *limit* determines the buffer size limit used by the returned :class:`Stream` instance. By default the limit is set to 64 KiB.
```

   .. versionadded:: 3.8

.. coroutinefunction:: connect_write_pipe(pipe, *, limit=2**16)

>    Takes a a :term:`file-like object <file object>` to return a

Reviewer: ```suggestion
   Takes a a :term:`file-like object <file object>` *pipe* to return a
```

   :class:`Stream` object of the mode ``WRITE`` that has similar API of
   :class:`StreamWriter`. It can also be used as an async context manager.

> 

Reviewer: ```
   *limit* determines the buffer size limit used by the returned :class:`Stream` instance. By default the limit is set to 64 KiB.
```

   .. versionadded:: 3.8

.. rubric:: Unix Sockets

.. function:: connect_unix(path=None, *, limit=2**16, ssl=None, \
                           sock=None, server_hostname=None, \
                           ssl_handshake_timeout=None)

>    Establish a Unix socket connection and return an awaitable :class:`Stream`

Reviewer: ```suggestion
   Establish a Unix socket connection to socket with *path* address and return an awaitable :class:`Stream`
```

   object of the mode ``READWRITE`` that can be used as a reader and a writer.
   The function can also be used as an async context manager where it's
   automatically awaited.

   Similar to :func:`connect` but operates on Unix sockets.

> 

Reviewer: Please describe the rest arguments: limit should be explained directly, all others can be mentioned by a link to `loop.create_unix_connection()`

Author: Added doc for limit like above functions and added a note that rest of arguments are passed to `loop.create_unix_connection()`. Thanks

   .. deprecated-removed:: 3.8 3.10

>       `open_unix_connection()` is deprecated if favor of :func:`connect_unix`.

Reviewer: ```suggestion
      ``open_unix_connection()`` is deprecated if favor of :func:`connect_unix`.
```

   .. deprecated-removed:: 3.8 3.10

      `start_unix_server()` is deprecated in favor of :class:`UnixStreamServer`.

---------

StreamServer
============

.. class:: StreamServer(client_connected_cb, /, host=None, port=None, *, \
                        limit=2**16, family=socket.AF_UNSPEC, \
                        flags=socket.AI_PASSIVE, sock=None, backlog=100, \
                        ssl=None, reuse_address=None, reuse_port=None, \
                        ssl_handshake_timeout=None, shutdown_timeout=60)

>    .. coroutinefunction:: start_serving

Reviewer: ```suggestion
   .. coroutinemethod:: start_serving()
```

      Binds to the given host and port to start the server. This method is

>       automatically called during ``__enter__`` when :class:`Stream` is

Reviewer: ```suggestion
      automatically called during ``__aenter__`` when :class:`Stream` is
```

>       used as a context manager.

Reviewer: ```suggestion
      used as an async context manager.
```

>    .. coroutinefunction:: close

Reviewer: ```suggestion
   .. coroutinemethod:: close()
```

      Closes the connection. This method is automatically called during

>       ``__exit__`` when :class:`StreamServer` is used as a context manager.

Reviewer: ```suggestion
      ``__aexit__`` when :class:`StreamServer` is used as an async context manager.
```

>    .. method:: is_serving

Reviewer: ```suggestion
   .. method:: is_serving()
```

   .. versionadded:: 3.8

UnixStreamServer
================

.. class:: UnixStreamServer(client_connected_cb, /, path=None, *, \
                            limit=2**16, sock=None, backlog=100, \
                            ssl=None, ssl_handshake_timeout=None, shutdown_timeout=60)

   .. coroutinefunction:: start_serving

      Binds to the given host and port to start the server. This method is
      automatically called during ``__enter__`` when :class:`Stream` is
      used as a context manager.

   .. coroutinefunction:: close

      Closes the connection. This method is automatically called during
      ``__exit__`` when :class:`UnixStreamServer` is used as a context manager.

   .. method:: is_serving

      Returns ``True`` if the server is bound and currently serving.

   .. attribute:: sockets

      Returns a tuple of sockets the server is bound to.

   .. availability:: Unix.

   .. versionadded:: 3.8

Stream
============

.. class:: Stream

   Represents a Stream object that provides APIs to read and write data
   to the IO stream . It includes the API provided by :class:`StreamReader`
   and :class:`StreamWriter`.

>    It is not recommended to instantiate *Stream* objects

Reviewer: `not recommended` sounds like "you shouldn't but you can".
But the stream *must* be instantiated by asyncio only.

I think the text should be
```
   Do not instantiate *Stream* objects
   directly; use API like :func:`connect` and :class:`StreamServer`
   instead.
```
I don't think we should enumerate all `connect_unix()` / `UnixStreamServer` / `connect_read_pipe()` / `connect_write_pipe()` here

Author: Thanks, I have added your suggestion.

   .. deprecated-removed:: 3.8 3.10

>       `start_unix_server()` is deprecated in favor of :class:`UnixStreamServer`.

Reviewer: ```suggestion
      ``start_unix_server()`` is deprecated in favor of :class:`UnixStreamServer`.
```

---------

StreamServer
============

.. class:: StreamServer(client_connected_cb, /, host=None, port=None, *, \
                        limit=2**16, family=socket.AF_UNSPEC, \
                        flags=socket.AI_PASSIVE, sock=None, backlog=100, \
                        ssl=None, reuse_address=None, reuse_port=None, \

>                         ssl_handshake_timeout=None, shutdown_timeout=60)

Reviewer: arguments should be described here. A link to `loop.create_server()` can be used to don't repeat `family`, `flags` etc.

Author: Added doc for client_connected_cb and limit. I have added a note that rest of the arguments are passed to `loop.create_server`.

      to be closed.

      This method can be called if the server is already accepting
      connections.  Only one ``serve_forever`` task can exist per
      one *Server* object.

   .. method:: is_serving()

      Returns ``True`` if the server is bound and currently serving.

   .. method:: bind()

      Bind the server to the given *host* and *port*. This method is
      automatically called during ``__aenter__`` when :class:`StreamServer` is
      used as an async context manager.

   .. method:: is_bound()

      Return ``True`` if the server is bound.

   .. coroutinemethod:: abort()

      Closes the connection and cancels all pending tasks.

   .. coroutinemethod:: close()

      Closes the connection. This method is automatically called during
      ``__aexit__`` when :class:`StreamServer` is used as an async context
      manager.

   .. attribute:: sockets

      Returns a tuple of socket objects the server is bound to.

   .. versionadded:: 3.8

UnixStreamServer
================

.. class:: UnixStreamServer(client_connected_cb, /, path=None, *, \
                            limit=2**16, sock=None, backlog=100, \

>                             ssl=None, ssl_handshake_timeout=None, shutdown_timeout=60)

Reviewer: Describe constructor arguments please

Author: Added doc for client_connected_cb and limit. I have added a note that rest of the arguments are passed to `loop.create_unix_server`.

TCP echo client using streams
-----------------------------

TCP echo client using the :func:`asyncio.connect` function::

    import asyncio

    async def tcp_echo_client(message):

>         stream = await asyncio.connect('127.0.0.1', 8888)

Reviewer: async context manager here?

        print(f'Send: {message!r}')

>         stream.write(message.encode())

Reviewer: ```suggestion
        await stream.write(message.encode())
```

TCP echo server using streams
-----------------------------

TCP echo server using the :class:`asyncio.StreamServer` class::

    import asyncio

    async def handle_echo(stream):
        data = await stream.read(100)
        message = data.decode()
        addr = stream.get_extra_info('peername')

        print(f"Received {message!r} from {addr!r}")

        print(f"Send: {message!r}")

>         stream.write(data)

Reviewer: ```suggestion
        await stream.write(data)
```
plus drop the following `await stream.drain()`.
`await stream.write()` implies write buffer draining.

        await stream.drain()

        print("Close the connection")
        await stream.close()

>     async def main():

Reviewer: ```
async def main():
    async with asyncio.StreamServer(
            handle_echo, '127.0.0.1', 8888) as server:

        await server.start_serving()
        addr = server.sockets[0].getsockname()
        print(f'Serving on {addr}')
        await server.serve_forever()
```

    async def print_http_headers(url):
        url = urllib.parse.urlsplit(url)
        if url.scheme == 'https':
            stream = await asyncio.connect(url.hostname, 443, ssl=True)
        else:
            stream = await asyncio.connect(url.hostname, 80)

        query = (
            f"HEAD {url.path or '/'} HTTP/1.0\r\n"
            f"Host: {url.hostname}\r\n"
            f"\r\n"
        )

>         stream.write(query.encode('latin-1'))

Reviewer: ```suggestion
        await stream.write(query.encode('latin-1'))
```

    async def print_http_headers(url):
        url = urllib.parse.urlsplit(url)
        if url.scheme == 'https':
            stream = await asyncio.connect(url.hostname, 443, ssl=True)
        else:
            stream = await asyncio.connect(url.hostname, 80)

        query = (
            f"HEAD {url.path or '/'} HTTP/1.0\r\n"
            f"Host: {url.hostname}\r\n"
            f"\r\n"
        )

        stream.write(query.encode('latin-1'))

>         while True:

Author: Some PEP 572 related example.

```python
while True:
    line = await stream.readline()
    if not line:
        break

    line = line.decode('latin1').rstrip()
    if line:
        print(f'HTTP header> {line}')
```

```python
while (line := await stream.readline()):
    line = line.decode('latin1').rstrip()
    if line:
        print(f'HTTP header> {line}')
```

Reviewer: Hmm. Agree, `:=` provides better readable code. Please apply.
